// angular2/ts/src/core/change_detection/directive_record.js
System.register(['angular2/src/facade/lang', './constants'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var lang_1, constants_1;
    var DirectiveIndex, DirectiveRecord;
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (constants_1_1) {
                constants_1 = constants_1_1;
            }],
        execute: function() {
            class DirectiveIndex {
                constructor(elementIndex, directiveIndex) {
                    this.elementIndex = elementIndex;
                    this.directiveIndex = directiveIndex;
                }
                get name() { return `${this.elementIndex}_${this.directiveIndex}`; }
            }
            exports_1("DirectiveIndex", DirectiveIndex);
            class DirectiveRecord {
                constructor({ directiveIndex, callAfterContentInit, callAfterContentChecked, callAfterViewInit, callAfterViewChecked, callOnChanges, callDoCheck, callOnInit, callOnDestroy, changeDetection, outputs } = {}) {
                    this.directiveIndex = directiveIndex;
                    this.callAfterContentInit = lang_1.normalizeBool(callAfterContentInit);
                    this.callAfterContentChecked = lang_1.normalizeBool(callAfterContentChecked);
                    this.callOnChanges = lang_1.normalizeBool(callOnChanges);
                    this.callAfterViewInit = lang_1.normalizeBool(callAfterViewInit);
                    this.callAfterViewChecked = lang_1.normalizeBool(callAfterViewChecked);
                    this.callDoCheck = lang_1.normalizeBool(callDoCheck);
                    this.callOnInit = lang_1.normalizeBool(callOnInit);
                    this.callOnDestroy = lang_1.normalizeBool(callOnDestroy);
                    this.changeDetection = changeDetection;
                    this.outputs = outputs;
                }
                isDefaultChangeDetection() {
                    return constants_1.isDefaultChangeDetectionStrategy(this.changeDetection);
                }
            }
            exports_1("DirectiveRecord", DirectiveRecord);
        }
    }
});
