// angular2/ts/src/core/change_detection/interfaces.js
System.register([], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var DebugContext, ChangeDetectorGenConfig, ChangeDetectorDefinition;
    return {
        setters:[],
        execute: function() {
            class DebugContext {
                constructor(element, componentElement, directive, context, locals, injector) {
                    this.element = element;
                    this.componentElement = componentElement;
                    this.directive = directive;
                    this.context = context;
                    this.locals = locals;
                    this.injector = injector;
                }
            }
            exports_1("DebugContext", DebugContext);
            class ChangeDetectorGenConfig {
                constructor(genDebugInfo, logBindingUpdate, useJit) {
                    this.genDebugInfo = genDebugInfo;
                    this.logBindingUpdate = logBindingUpdate;
                    this.useJit = useJit;
                }
            }
            exports_1("ChangeDetectorGenConfig", ChangeDetectorGenConfig);
            class ChangeDetectorDefinition {
                constructor(id, strategy, variableNames, bindingRecords, eventRecords, directiveRecords, genConfig) {
                    this.id = id;
                    this.strategy = strategy;
                    this.variableNames = variableNames;
                    this.bindingRecords = bindingRecords;
                    this.eventRecords = eventRecords;
                    this.directiveRecords = directiveRecords;
                    this.genConfig = genConfig;
                }
            }
            exports_1("ChangeDetectorDefinition", ChangeDetectorDefinition);
        }
    }
});
