// angular2/ts/src/router/route_config_impl.js
System.register(['angular2/src/facade/lang'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var lang_1;
    var RouteConfig, Route, AuxRoute, AsyncRoute, Redirect;
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            }],
        execute: function() {
            /**
             * The `RouteConfig` decorator defines routes for a given component.
             *
             * It takes an array of {@link RouteDefinition}s.
             */
            let RouteConfig = class RouteConfig {
                constructor(configs) {
                    this.configs = configs;
                }
            };
            RouteConfig = __decorate([
                lang_1.CONST(), 
                __metadata('design:paramtypes', [Array])
            ], RouteConfig);
            exports_1("RouteConfig", RouteConfig);
            /**
             * `Route` is a type of {@link RouteDefinition} used to route a path to a component.
             *
             * It has the following properties:
             * - `path` is a string that uses the route matcher DSL.
             * - `component` a component type.
             * - `name` is an optional `CamelCase` string representing the name of the route.
             * - `data` is an optional property of any type representing arbitrary route metadata for the given
             * route. It is injectable via {@link RouteData}.
             * - `useAsDefault` is a boolean value. If `true`, the child route will be navigated to if no child
             * route is specified during the navigation.
             *
             * ### Example
             * ```
             * import {RouteConfig, Route} from 'angular2/router';
             *
             * @RouteConfig([
             *   new Route({path: '/home', component: HomeCmp, name: 'HomeCmp' })
             * ])
             * class MyApp {}
             * ```
             */
            let Route = class Route {
                constructor({ path, component, name, data, useAsDefault }) {
                    // added next three properties to work around https://github.com/Microsoft/TypeScript/issues/4107
                    this.aux = null;
                    this.loader = null;
                    this.redirectTo = null;
                    this.path = path;
                    this.component = component;
                    this.name = name;
                    this.data = data;
                    this.useAsDefault = useAsDefault;
                }
            };
            Route = __decorate([
                lang_1.CONST(), 
                __metadata('design:paramtypes', [Object])
            ], Route);
            exports_1("Route", Route);
            /**
             * `AuxRoute` is a type of {@link RouteDefinition} used to define an auxiliary route.
             *
             * It takes an object with the following properties:
             * - `path` is a string that uses the route matcher DSL.
             * - `component` a component type.
             * - `name` is an optional `CamelCase` string representing the name of the route.
             * - `data` is an optional property of any type representing arbitrary route metadata for the given
             * route. It is injectable via {@link RouteData}.
             *
             * ### Example
             * ```
             * import {RouteConfig, AuxRoute} from 'angular2/router';
             *
             * @RouteConfig([
             *   new AuxRoute({path: '/home', component: HomeCmp})
             * ])
             * class MyApp {}
             * ```
             */
            let AuxRoute = class AuxRoute {
                constructor({ path, component, name }) {
                    this.data = null;
                    // added next three properties to work around https://github.com/Microsoft/TypeScript/issues/4107
                    this.aux = null;
                    this.loader = null;
                    this.redirectTo = null;
                    this.useAsDefault = false;
                    this.path = path;
                    this.component = component;
                    this.name = name;
                }
            };
            AuxRoute = __decorate([
                lang_1.CONST(), 
                __metadata('design:paramtypes', [Object])
            ], AuxRoute);
            exports_1("AuxRoute", AuxRoute);
            /**
             * `AsyncRoute` is a type of {@link RouteDefinition} used to route a path to an asynchronously
             * loaded component.
             *
             * It has the following properties:
             * - `path` is a string that uses the route matcher DSL.
             * - `loader` is a function that returns a promise that resolves to a component.
             * - `name` is an optional `CamelCase` string representing the name of the route.
             * - `data` is an optional property of any type representing arbitrary route metadata for the given
             * route. It is injectable via {@link RouteData}.
             * - `useAsDefault` is a boolean value. If `true`, the child route will be navigated to if no child
             * route is specified during the navigation.
             *
             * ### Example
             * ```
             * import {RouteConfig, AsyncRoute} from 'angular2/router';
             *
             * @RouteConfig([
             *   new AsyncRoute({path: '/home', loader: () => Promise.resolve(MyLoadedCmp), name:
             * 'MyLoadedCmp'})
             * ])
             * class MyApp {}
             * ```
             */
            let AsyncRoute = class AsyncRoute {
                constructor({ path, loader, name, data, useAsDefault }) {
                    this.aux = null;
                    this.path = path;
                    this.loader = loader;
                    this.name = name;
                    this.data = data;
                    this.useAsDefault = useAsDefault;
                }
            };
            AsyncRoute = __decorate([
                lang_1.CONST(), 
                __metadata('design:paramtypes', [Object])
            ], AsyncRoute);
            exports_1("AsyncRoute", AsyncRoute);
            /**
             * `Redirect` is a type of {@link RouteDefinition} used to route a path to a canonical route.
             *
             * It has the following properties:
             * - `path` is a string that uses the route matcher DSL.
             * - `redirectTo` is an array representing the link DSL.
             *
             * Note that redirects **do not** affect how links are generated. For that, see the `useAsDefault`
             * option.
             *
             * ### Example
             * ```
             * import {RouteConfig, Route, Redirect} from 'angular2/router';
             *
             * @RouteConfig([
             *   new Redirect({path: '/', redirectTo: ['/Home'] }),
             *   new Route({path: '/home', component: HomeCmp, name: 'Home'})
             * ])
             * class MyApp {}
             * ```
             */
            let Redirect = class Redirect {
                constructor({ path, redirectTo }) {
                    this.name = null;
                    // added next three properties to work around https://github.com/Microsoft/TypeScript/issues/4107
                    this.loader = null;
                    this.data = null;
                    this.aux = null;
                    this.useAsDefault = false;
                    this.path = path;
                    this.redirectTo = redirectTo;
                }
            };
            Redirect = __decorate([
                lang_1.CONST(), 
                __metadata('design:paramtypes', [Object])
            ], Redirect);
            exports_1("Redirect", Redirect);
        }
    }
});
