// reflect-metadata/test/reflect/reflect-getmetadatakeys.js
// Reflect.getMetadataKeys ( target [, propertyKey] )
// - https://github.com/jonathandturner/decorators/blob/master/specs/metadata.md#reflectgetmetadatakeys--target--propertykey-
System.register(["../../Reflect", "assert"], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var assert;
    function ReflectGetMetadataKeysInvalidTarget() {
        // 1. If Type(target) is not Object, throw a TypeError exception.
        assert.throws(() => Reflect.getMetadataKeys(undefined, undefined), TypeError);
    }
    exports_1("ReflectGetMetadataKeysInvalidTarget", ReflectGetMetadataKeysInvalidTarget);
    function ReflectGetMetadataKeysWithoutTargetKeyWhenNotDefined() {
        let obj = {};
        let result = Reflect.getMetadataKeys(obj, undefined);
        assert.deepEqual(result, []);
    }
    exports_1("ReflectGetMetadataKeysWithoutTargetKeyWhenNotDefined", ReflectGetMetadataKeysWithoutTargetKeyWhenNotDefined);
    function ReflectGetMetadataKeysWithoutTargetKeyWhenDefined() {
        let obj = {};
        Reflect.defineMetadata("key", "value", obj, undefined);
        let result = Reflect.getMetadataKeys(obj, undefined);
        assert.deepEqual(result, ["key"]);
    }
    exports_1("ReflectGetMetadataKeysWithoutTargetKeyWhenDefined", ReflectGetMetadataKeysWithoutTargetKeyWhenDefined);
    function ReflectGetMetadataKeysWithoutTargetKeyWhenDefinedOnPrototype() {
        let prototype = {};
        let obj = Object.create(prototype);
        Reflect.defineMetadata("key", "value", prototype, undefined);
        let result = Reflect.getMetadataKeys(obj, undefined);
        assert.deepEqual(result, ["key"]);
    }
    exports_1("ReflectGetMetadataKeysWithoutTargetKeyWhenDefinedOnPrototype", ReflectGetMetadataKeysWithoutTargetKeyWhenDefinedOnPrototype);
    function ReflectGetMetadataKeysOrderWithoutTargetKey() {
        let obj = {};
        Reflect.defineMetadata("key1", "value", obj, undefined);
        Reflect.defineMetadata("key0", "value", obj, undefined);
        let result = Reflect.getMetadataKeys(obj, undefined);
        assert.deepEqual(result, ["key1", "key0"]);
    }
    exports_1("ReflectGetMetadataKeysOrderWithoutTargetKey", ReflectGetMetadataKeysOrderWithoutTargetKey);
    function ReflectGetMetadataKeysOrderAfterRedefineWithoutTargetKey() {
        let obj = {};
        Reflect.defineMetadata("key1", "value", obj, undefined);
        Reflect.defineMetadata("key0", "value", obj, undefined);
        Reflect.defineMetadata("key1", "value", obj, undefined);
        let result = Reflect.getMetadataKeys(obj, undefined);
        assert.deepEqual(result, ["key1", "key0"]);
    }
    exports_1("ReflectGetMetadataKeysOrderAfterRedefineWithoutTargetKey", ReflectGetMetadataKeysOrderAfterRedefineWithoutTargetKey);
    function ReflectGetMetadataKeysOrderWithoutTargetKeyWhenDefinedOnPrototype() {
        let prototype = {};
        Reflect.defineMetadata("key2", "value", prototype, undefined);
        let obj = Object.create(prototype);
        Reflect.defineMetadata("key1", "value", obj, undefined);
        Reflect.defineMetadata("key0", "value", obj, undefined);
        let result = Reflect.getMetadataKeys(obj, undefined);
        assert.deepEqual(result, ["key1", "key0", "key2"]);
    }
    exports_1("ReflectGetMetadataKeysOrderWithoutTargetKeyWhenDefinedOnPrototype", ReflectGetMetadataKeysOrderWithoutTargetKeyWhenDefinedOnPrototype);
    function ReflectGetMetadataKeysWithTargetKeyWhenNotDefined() {
        let obj = {};
        let result = Reflect.getMetadataKeys(obj, "name");
        assert.deepEqual(result, []);
    }
    exports_1("ReflectGetMetadataKeysWithTargetKeyWhenNotDefined", ReflectGetMetadataKeysWithTargetKeyWhenNotDefined);
    function ReflectGetMetadataKeysWithTargetKeyWhenDefined() {
        let obj = {};
        Reflect.defineMetadata("key", "value", obj, "name");
        let result = Reflect.getMetadataKeys(obj, "name");
        assert.deepEqual(result, ["key"]);
    }
    exports_1("ReflectGetMetadataKeysWithTargetKeyWhenDefined", ReflectGetMetadataKeysWithTargetKeyWhenDefined);
    function ReflectGetMetadataKeysWithTargetKeyWhenDefinedOnPrototype() {
        let prototype = {};
        let obj = Object.create(prototype);
        Reflect.defineMetadata("key", "value", prototype, "name");
        let result = Reflect.getMetadataKeys(obj, "name");
        assert.deepEqual(result, ["key"]);
    }
    exports_1("ReflectGetMetadataKeysWithTargetKeyWhenDefinedOnPrototype", ReflectGetMetadataKeysWithTargetKeyWhenDefinedOnPrototype);
    function ReflectGetMetadataKeysOrderAfterRedefineWithTargetKey() {
        let obj = {};
        Reflect.defineMetadata("key1", "value", obj, "name");
        Reflect.defineMetadata("key0", "value", obj, "name");
        Reflect.defineMetadata("key1", "value", obj, "name");
        let result = Reflect.getMetadataKeys(obj, "name");
        assert.deepEqual(result, ["key1", "key0"]);
    }
    exports_1("ReflectGetMetadataKeysOrderAfterRedefineWithTargetKey", ReflectGetMetadataKeysOrderAfterRedefineWithTargetKey);
    function ReflectGetMetadataKeysOrderWithTargetKeyWhenDefinedOnPrototype() {
        let prototype = {};
        Reflect.defineMetadata("key2", "value", prototype, "name");
        let obj = Object.create(prototype);
        Reflect.defineMetadata("key1", "value", obj, "name");
        Reflect.defineMetadata("key0", "value", obj, "name");
        let result = Reflect.getMetadataKeys(obj, "name");
        assert.deepEqual(result, ["key1", "key0", "key2"]);
    }
    exports_1("ReflectGetMetadataKeysOrderWithTargetKeyWhenDefinedOnPrototype", ReflectGetMetadataKeysOrderWithTargetKeyWhenDefinedOnPrototype);
    return {
        setters:[
            function (_1) {},
            function (assert_1) {
                assert = assert_1;
            }],
        execute: function() {
        }
    }
});
