// reflect-metadata/test/reflect/reflect-getownmetadata.js
// Reflect.getOwnMetadata ( metadataKey, target [, propertyKey] )
// - https://github.com/jonathandturner/decorators/blob/master/specs/metadata.md#reflectgetownmetadata--metadatakey-target--propertykey-
System.register(["../../Reflect", "assert"], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var assert;
    function ReflectGetOwnMetadataInvalidTarget() {
        assert.throws(() => Reflect.getOwnMetadata("key", undefined, undefined), TypeError);
    }
    exports_1("ReflectGetOwnMetadataInvalidTarget", ReflectGetOwnMetadataInvalidTarget);
    function ReflectGetOwnMetadataWithoutTargetKeyWhenNotDefined() {
        let obj = {};
        let result = Reflect.getOwnMetadata("key", obj, undefined);
        assert.equal(result, undefined);
    }
    exports_1("ReflectGetOwnMetadataWithoutTargetKeyWhenNotDefined", ReflectGetOwnMetadataWithoutTargetKeyWhenNotDefined);
    function ReflectGetOwnMetadataWithoutTargetKeyWhenDefined() {
        let obj = {};
        Reflect.defineMetadata("key", "value", obj, undefined);
        let result = Reflect.getOwnMetadata("key", obj, undefined);
        assert.equal(result, "value");
    }
    exports_1("ReflectGetOwnMetadataWithoutTargetKeyWhenDefined", ReflectGetOwnMetadataWithoutTargetKeyWhenDefined);
    function ReflectGetOwnMetadataWithoutTargetKeyWhenDefinedOnPrototype() {
        let prototype = {};
        let obj = Object.create(prototype);
        Reflect.defineMetadata("key", "value", prototype, undefined);
        let result = Reflect.getOwnMetadata("key", obj, undefined);
        assert.equal(result, undefined);
    }
    exports_1("ReflectGetOwnMetadataWithoutTargetKeyWhenDefinedOnPrototype", ReflectGetOwnMetadataWithoutTargetKeyWhenDefinedOnPrototype);
    function ReflectGetOwnMetadataWithTargetKeyWhenNotDefined() {
        let obj = {};
        let result = Reflect.getOwnMetadata("key", obj, "name");
        assert.equal(result, undefined);
    }
    exports_1("ReflectGetOwnMetadataWithTargetKeyWhenNotDefined", ReflectGetOwnMetadataWithTargetKeyWhenNotDefined);
    function ReflectGetOwnMetadataWithTargetKeyWhenDefined() {
        let obj = {};
        Reflect.defineMetadata("key", "value", obj, "name");
        let result = Reflect.getOwnMetadata("key", obj, "name");
        assert.equal(result, "value");
    }
    exports_1("ReflectGetOwnMetadataWithTargetKeyWhenDefined", ReflectGetOwnMetadataWithTargetKeyWhenDefined);
    function ReflectGetOwnMetadataWithTargetKeyWhenDefinedOnPrototype() {
        let prototype = {};
        let obj = Object.create(prototype);
        Reflect.defineMetadata("key", "value", prototype, "name");
        let result = Reflect.getOwnMetadata("key", obj, "name");
        assert.equal(result, undefined);
    }
    exports_1("ReflectGetOwnMetadataWithTargetKeyWhenDefinedOnPrototype", ReflectGetOwnMetadataWithTargetKeyWhenDefinedOnPrototype);
    return {
        setters:[
            function (_1) {},
            function (assert_1) {
                assert = assert_1;
            }],
        execute: function() {
        }
    }
});
