// angular2/ts/src/common/forms/directives/ng_control_group.js
System.register(['angular2/core', 'angular2/src/facade/lang', './control_container', './shared', '../validators'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __param = (this && this.__param) || function (paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    };
    var core_1, lang_1, control_container_1, shared_1, validators_1;
    var controlGroupProvider, NgControlGroup;
    return {
        setters:[
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (control_container_1_1) {
                control_container_1 = control_container_1_1;
            },
            function (shared_1_1) {
                shared_1 = shared_1_1;
            },
            function (validators_1_1) {
                validators_1 = validators_1_1;
            }],
        execute: function() {
            controlGroupProvider = lang_1.CONST_EXPR(new core_1.Provider(control_container_1.ControlContainer, { useExisting: core_1.forwardRef(() => NgControlGroup) }));
            /**
             * Creates and binds a control group to a DOM element.
             *
             * This directive can only be used as a child of {@link NgForm} or {@link NgFormModel}.
             *
             * ### Example ([live demo](http://plnkr.co/edit/7EJ11uGeaggViYM6T5nq?p=preview))
             *
             * ```typescript
             * @Component({
             *   selector: 'my-app',
             *   directives: [FORM_DIRECTIVES],
             * })
             * @View({
             *   template: `
             *     <div>
             *       <h2>Angular2 Control &amp; ControlGroup Example</h2>
             *       <form #f="ngForm">
             *         <div ngControlGroup="name" #cg-name="form">
             *           <h3>Enter your name:</h3>
             *           <p>First: <input ngControl="first" required></p>
             *           <p>Middle: <input ngControl="middle"></p>
             *           <p>Last: <input ngControl="last" required></p>
             *         </div>
             *         <h3>Name value:</h3>
             *         <pre>{{valueOf(cgName)}}</pre>
             *         <p>Name is {{cgName?.control?.valid ? "valid" : "invalid"}}</p>
             *         <h3>What's your favorite food?</h3>
             *         <p><input ngControl="food"></p>
             *         <h3>Form value</h3>
             *         <pre>{{valueOf(f)}}</pre>
             *       </form>
             *     </div>
             *   `,
             *   directives: [FORM_DIRECTIVES]
             * })
             * export class App {
             *   valueOf(cg: NgControlGroup): string {
             *     if (cg.control == null) {
             *       return null;
             *     }
             *     return JSON.stringify(cg.control.value, null, 2);
             *   }
             * }
             * ```
             *
             * This example declares a control group for a user's name. The value and validation state of
             * this group can be accessed separately from the overall form.
             */
            let NgControlGroup = class NgControlGroup extends control_container_1.ControlContainer {
                constructor(parent, _validators, _asyncValidators) {
                    super();
                    this._validators = _validators;
                    this._asyncValidators = _asyncValidators;
                    this._parent = parent;
                }
                ngOnInit() { this.formDirective.addControlGroup(this); }
                ngOnDestroy() { this.formDirective.removeControlGroup(this); }
                /**
                 * Get the {@link ControlGroup} backing this binding.
                 */
                get control() { return this.formDirective.getControlGroup(this); }
                /**
                 * Get the path to this control group.
                 */
                get path() { return shared_1.controlPath(this.name, this._parent); }
                /**
                 * Get the {@link Form} to which this group belongs.
                 */
                get formDirective() { return this._parent.formDirective; }
                get validator() { return shared_1.composeValidators(this._validators); }
                get asyncValidator() { return shared_1.composeAsyncValidators(this._asyncValidators); }
            };
            NgControlGroup = __decorate([
                core_1.Directive({
                    selector: '[ngControlGroup]',
                    providers: [controlGroupProvider],
                    inputs: ['name: ngControlGroup'],
                    exportAs: 'ngForm'
                }),
                __param(0, core_1.Host()),
                __param(0, core_1.SkipSelf()),
                __param(1, core_1.Optional()),
                __param(1, core_1.Self()),
                __param(1, core_1.Inject(validators_1.NG_VALIDATORS)),
                __param(2, core_1.Optional()),
                __param(2, core_1.Self()),
                __param(2, core_1.Inject(validators_1.NG_ASYNC_VALIDATORS)), 
                __metadata('design:paramtypes', [control_container_1.ControlContainer, Array, Array])
            ], NgControlGroup);
            exports_1("NgControlGroup", NgControlGroup);
        }
    }
});
