// angular2/ts/src/common/forms/directives/ng_control_name.js
System.register(['angular2/src/facade/lang', 'angular2/src/facade/async', 'angular2/core', './control_container', './ng_control', './control_value_accessor', './shared', '../validators'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __param = (this && this.__param) || function (paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    };
    var lang_1, async_1, core_1, control_container_1, ng_control_1, control_value_accessor_1, shared_1, validators_1;
    var controlNameBinding, NgControlName;
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (async_1_1) {
                async_1 = async_1_1;
            },
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (control_container_1_1) {
                control_container_1 = control_container_1_1;
            },
            function (ng_control_1_1) {
                ng_control_1 = ng_control_1_1;
            },
            function (control_value_accessor_1_1) {
                control_value_accessor_1 = control_value_accessor_1_1;
            },
            function (shared_1_1) {
                shared_1 = shared_1_1;
            },
            function (validators_1_1) {
                validators_1 = validators_1_1;
            }],
        execute: function() {
            controlNameBinding = lang_1.CONST_EXPR(new core_1.Provider(ng_control_1.NgControl, { useExisting: core_1.forwardRef(() => NgControlName) }));
            /**
             * Creates and binds a control with a specified name to a DOM element.
             *
             * This directive can only be used as a child of {@link NgForm} or {@link NgFormModel}.
            
             * ### Example
             *
             * In this example, we create the login and password controls.
             * We can work with each control separately: check its validity, get its value, listen to its
             * changes.
             *
             *  ```
             * @Component({
             *      selector: "login-comp",
             *      directives: [FORM_DIRECTIVES],
             *      template: `
             *        <form #f="ngForm" (submit)='onLogIn(f.value)'>
             *          Login <input type='text' ngControl='login' #l="form">
             *          <div *ngIf="!l.valid">Login is invalid</div>
             *
             *          Password <input type='password' ngControl='password'>
             *          <button type='submit'>Log in!</button>
             *        </form>
             *      `})
             * class LoginComp {
             *  onLogIn(value): void {
             *    // value === {login: 'some login', password: 'some password'}
             *  }
             * }
             *  ```
             *
             * We can also use ngModel to bind a domain model to the form.
             *
             *  ```
             * @Component({
             *      selector: "login-comp",
             *      directives: [FORM_DIRECTIVES],
             *      template: `
             *        <form (submit)='onLogIn()'>
             *          Login <input type='text' ngControl='login' [(ngModel)]="credentials.login">
             *          Password <input type='password' ngControl='password'
             *                          [(ngModel)]="credentials.password">
             *          <button type='submit'>Log in!</button>
             *        </form>
             *      `})
             * class LoginComp {
             *  credentials: {login:string, password:string};
             *
             *  onLogIn(): void {
             *    // this.credentials.login === "some login"
             *    // this.credentials.password === "some password"
             *  }
             * }
             *  ```
             */
            let NgControlName = class NgControlName extends ng_control_1.NgControl {
                constructor(_parent, _validators, _asyncValidators, valueAccessors) {
                    super();
                    this._parent = _parent;
                    this._validators = _validators;
                    this._asyncValidators = _asyncValidators;
                    /** @internal */
                    this.update = new async_1.EventEmitter();
                    this._added = false;
                    this.valueAccessor = shared_1.selectValueAccessor(this, valueAccessors);
                }
                ngOnChanges(changes) {
                    if (!this._added) {
                        this.formDirective.addControl(this);
                        this._added = true;
                    }
                    if (shared_1.isPropertyUpdated(changes, this.viewModel)) {
                        this.viewModel = this.model;
                        this.formDirective.updateModel(this, this.model);
                    }
                }
                ngOnDestroy() { this.formDirective.removeControl(this); }
                viewToModelUpdate(newValue) {
                    this.viewModel = newValue;
                    async_1.ObservableWrapper.callEmit(this.update, newValue);
                }
                get path() { return shared_1.controlPath(this.name, this._parent); }
                get formDirective() { return this._parent.formDirective; }
                get validator() { return shared_1.composeValidators(this._validators); }
                get asyncValidator() { return shared_1.composeAsyncValidators(this._asyncValidators); }
                get control() { return this.formDirective.getControl(this); }
            };
            NgControlName = __decorate([
                core_1.Directive({
                    selector: '[ngControl]',
                    bindings: [controlNameBinding],
                    inputs: ['name: ngControl', 'model: ngModel'],
                    outputs: ['update: ngModelChange'],
                    exportAs: 'ngForm'
                }),
                __param(0, core_1.Host()),
                __param(0, core_1.SkipSelf()),
                __param(1, core_1.Optional()),
                __param(1, core_1.Self()),
                __param(1, core_1.Inject(validators_1.NG_VALIDATORS)),
                __param(2, core_1.Optional()),
                __param(2, core_1.Self()),
                __param(2, core_1.Inject(validators_1.NG_ASYNC_VALIDATORS)),
                __param(3, core_1.Optional()),
                __param(3, core_1.Self()),
                __param(3, core_1.Inject(control_value_accessor_1.NG_VALUE_ACCESSOR)), 
                __metadata('design:paramtypes', [control_container_1.ControlContainer, Array, Array, Array])
            ], NgControlName);
            exports_1("NgControlName", NgControlName);
        }
    }
});
