// angular2/ts/src/common/forms/directives/ng_form_model.js
System.register(['angular2/src/facade/lang', 'angular2/src/facade/collection', 'angular2/src/facade/async', 'angular2/core', './control_container', './shared', '../validators'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __param = (this && this.__param) || function (paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    };
    var lang_1, collection_1, async_1, core_1, control_container_1, shared_1, validators_1;
    var formDirectiveProvider, NgFormModel;
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (collection_1_1) {
                collection_1 = collection_1_1;
            },
            function (async_1_1) {
                async_1 = async_1_1;
            },
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (control_container_1_1) {
                control_container_1 = control_container_1_1;
            },
            function (shared_1_1) {
                shared_1 = shared_1_1;
            },
            function (validators_1_1) {
                validators_1 = validators_1_1;
            }],
        execute: function() {
            formDirectiveProvider = lang_1.CONST_EXPR(new core_1.Provider(control_container_1.ControlContainer, { useExisting: core_1.forwardRef(() => NgFormModel) }));
            /**
             * Binds an existing control group to a DOM element.
             *
             * ### Example ([live demo](http://plnkr.co/edit/jqrVirudY8anJxTMUjTP?p=preview))
             *
             * In this example, we bind the control group to the form element, and we bind the login and
             * password controls to the login and password elements.
             *
             *  ```typescript
             * @Component({
             *   selector: 'my-app',
             *   template: `
             *     <div>
             *       <h2>NgFormModel Example</h2>
             *       <form [ngFormModel]="loginForm">
             *         <p>Login: <input type="text" ngControl="login"></p>
             *         <p>Password: <input type="password" ngControl="password"></p>
             *       </form>
             *       <p>Value:</p>
             *       <pre>{{value}}</pre>
             *     </div>
             *   `,
             *   directives: [FORM_DIRECTIVES]
             * })
             * export class App {
             *   loginForm: ControlGroup;
             *
             *   constructor() {
             *     this.loginForm = new ControlGroup({
             *       login: new Control(""),
             *       password: new Control("")
             *     });
             *   }
             *
             *   get value(): string {
             *     return JSON.stringify(this.loginForm.value, null, 2);
             *   }
             * }
             *  ```
             *
             * We can also use ngModel to bind a domain model to the form.
             *
             *  ```typescript
             * @Component({
             *      selector: "login-comp",
             *      directives: [FORM_DIRECTIVES],
             *      template: `
             *        <form [ngFormModel]='loginForm'>
             *          Login <input type='text' ngControl='login' [(ngModel)]='credentials.login'>
             *          Password <input type='password' ngControl='password'
             *                          [(ngModel)]='credentials.password'>
             *          <button (click)="onLogin()">Login</button>
             *        </form>`
             *      })
             * class LoginComp {
             *  credentials: {login: string, password: string};
             *  loginForm: ControlGroup;
             *
             *  constructor() {
             *    this.loginForm = new ControlGroup({
             *      login: new Control(""),
             *      password: new Control("")
             *    });
             *  }
             *
             *  onLogin(): void {
             *    // this.credentials.login === 'some login'
             *    // this.credentials.password === 'some password'
             *  }
             * }
             *  ```
             */
            let NgFormModel = class NgFormModel extends control_container_1.ControlContainer {
                constructor(_validators, _asyncValidators) {
                    super();
                    this._validators = _validators;
                    this._asyncValidators = _asyncValidators;
                    this.form = null;
                    this.directives = [];
                    this.ngSubmit = new async_1.EventEmitter();
                }
                ngOnChanges(changes) {
                    if (collection_1.StringMapWrapper.contains(changes, "form")) {
                        var sync = shared_1.composeValidators(this._validators);
                        this.form.validator = validators_1.Validators.compose([this.form.validator, sync]);
                        var async = shared_1.composeAsyncValidators(this._asyncValidators);
                        this.form.asyncValidator = validators_1.Validators.composeAsync([this.form.asyncValidator, async]);
                        this.form.updateValueAndValidity({ onlySelf: true, emitEvent: false });
                    }
                    this._updateDomValue();
                }
                get formDirective() { return this; }
                get control() { return this.form; }
                get path() { return []; }
                addControl(dir) {
                    var ctrl = this.form.find(dir.path);
                    shared_1.setUpControl(ctrl, dir);
                    ctrl.updateValueAndValidity({ emitEvent: false });
                    this.directives.push(dir);
                }
                getControl(dir) { return this.form.find(dir.path); }
                removeControl(dir) { collection_1.ListWrapper.remove(this.directives, dir); }
                addControlGroup(dir) {
                    var ctrl = this.form.find(dir.path);
                    shared_1.setUpControlGroup(ctrl, dir);
                    ctrl.updateValueAndValidity({ emitEvent: false });
                }
                removeControlGroup(dir) { }
                getControlGroup(dir) {
                    return this.form.find(dir.path);
                }
                updateModel(dir, value) {
                    var ctrl = this.form.find(dir.path);
                    ctrl.updateValue(value);
                }
                onSubmit() {
                    async_1.ObservableWrapper.callEmit(this.ngSubmit, null);
                    return false;
                }
                /** @internal */
                _updateDomValue() {
                    this.directives.forEach(dir => {
                        var ctrl = this.form.find(dir.path);
                        dir.valueAccessor.writeValue(ctrl.value);
                    });
                }
            };
            NgFormModel = __decorate([
                core_1.Directive({
                    selector: '[ngFormModel]',
                    bindings: [formDirectiveProvider],
                    inputs: ['form: ngFormModel'],
                    host: { '(submit)': 'onSubmit()' },
                    outputs: ['ngSubmit'],
                    exportAs: 'ngForm'
                }),
                __param(0, core_1.Optional()),
                __param(0, core_1.Self()),
                __param(0, core_1.Inject(validators_1.NG_VALIDATORS)),
                __param(1, core_1.Optional()),
                __param(1, core_1.Self()),
                __param(1, core_1.Inject(validators_1.NG_ASYNC_VALIDATORS)), 
                __metadata('design:paramtypes', [Array, Array])
            ], NgFormModel);
            exports_1("NgFormModel", NgFormModel);
        }
    }
});
