// angular2/ts/src/core/debug/debug_renderer.js
System.register(['angular2/src/facade/lang', 'angular2/src/core/debug/debug_node'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var lang_1, debug_node_1;
    var DebugDomRootRenderer, DebugDomRenderer;
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (debug_node_1_1) {
                debug_node_1 = debug_node_1_1;
            }],
        execute: function() {
            class DebugDomRootRenderer {
                constructor(_delegate) {
                    this._delegate = _delegate;
                }
                renderComponent(componentProto) {
                    return new DebugDomRenderer(this, this._delegate.renderComponent(componentProto));
                }
            }
            exports_1("DebugDomRootRenderer", DebugDomRootRenderer);
            class DebugDomRenderer {
                constructor(_rootRenderer, _delegate) {
                    this._rootRenderer = _rootRenderer;
                    this._delegate = _delegate;
                }
                renderComponent(componentType) {
                    return this._rootRenderer.renderComponent(componentType);
                }
                selectRootElement(selector) {
                    var nativeEl = this._delegate.selectRootElement(selector);
                    var debugEl = new debug_node_1.DebugElement(nativeEl, null);
                    debug_node_1.indexDebugNode(debugEl);
                    return nativeEl;
                }
                createElement(parentElement, name) {
                    var nativeEl = this._delegate.createElement(parentElement, name);
                    var debugEl = new debug_node_1.DebugElement(nativeEl, debug_node_1.getDebugNode(parentElement));
                    debugEl.name = name;
                    debug_node_1.indexDebugNode(debugEl);
                    return nativeEl;
                }
                createViewRoot(hostElement) { return this._delegate.createViewRoot(hostElement); }
                createTemplateAnchor(parentElement) {
                    var comment = this._delegate.createTemplateAnchor(parentElement);
                    var debugEl = new debug_node_1.DebugNode(comment, debug_node_1.getDebugNode(parentElement));
                    debug_node_1.indexDebugNode(debugEl);
                    return comment;
                }
                createText(parentElement, value) {
                    var text = this._delegate.createText(parentElement, value);
                    var debugEl = new debug_node_1.DebugNode(text, debug_node_1.getDebugNode(parentElement));
                    debug_node_1.indexDebugNode(debugEl);
                    return text;
                }
                projectNodes(parentElement, nodes) {
                    var debugParent = debug_node_1.getDebugNode(parentElement);
                    if (lang_1.isPresent(debugParent) && debugParent instanceof debug_node_1.DebugElement) {
                        nodes.forEach((node) => { debugParent.addChild(debug_node_1.getDebugNode(node)); });
                    }
                    return this._delegate.projectNodes(parentElement, nodes);
                }
                attachViewAfter(node, viewRootNodes) {
                    var debugNode = debug_node_1.getDebugNode(node);
                    if (lang_1.isPresent(debugNode)) {
                        var debugParent = debugNode.parent;
                        if (viewRootNodes.length > 0 && lang_1.isPresent(debugParent)) {
                            var debugViewRootNodes = [];
                            viewRootNodes.forEach((rootNode) => debugViewRootNodes.push(debug_node_1.getDebugNode(rootNode)));
                            debugParent.insertChildrenAfter(debugNode, debugViewRootNodes);
                        }
                    }
                    return this._delegate.attachViewAfter(node, viewRootNodes);
                }
                detachView(viewRootNodes) {
                    viewRootNodes.forEach((node) => {
                        var debugNode = debug_node_1.getDebugNode(node);
                        if (lang_1.isPresent(debugNode) && lang_1.isPresent(debugNode.parent)) {
                            debugNode.parent.removeChild(debugNode);
                        }
                    });
                    return this._delegate.detachView(viewRootNodes);
                }
                destroyView(hostElement, viewAllNodes) {
                    viewAllNodes.forEach((node) => { debug_node_1.removeDebugNodeFromIndex(debug_node_1.getDebugNode(node)); });
                    return this._delegate.destroyView(hostElement, viewAllNodes);
                }
                listen(renderElement, name, callback) {
                    var debugEl = debug_node_1.getDebugNode(renderElement);
                    if (lang_1.isPresent(debugEl)) {
                        debugEl.listeners.push(new debug_node_1.EventListener(name, callback));
                    }
                    return this._delegate.listen(renderElement, name, callback);
                }
                listenGlobal(target, name, callback) {
                    return this._delegate.listenGlobal(target, name, callback);
                }
                setElementProperty(renderElement, propertyName, propertyValue) {
                    var debugEl = debug_node_1.getDebugNode(renderElement);
                    if (lang_1.isPresent(debugEl) && debugEl instanceof debug_node_1.DebugElement) {
                        debugEl.properties.set(propertyName, propertyValue);
                    }
                    return this._delegate.setElementProperty(renderElement, propertyName, propertyValue);
                }
                setElementAttribute(renderElement, attributeName, attributeValue) {
                    var debugEl = debug_node_1.getDebugNode(renderElement);
                    if (lang_1.isPresent(debugEl) && debugEl instanceof debug_node_1.DebugElement) {
                        debugEl.attributes.set(attributeName, attributeValue);
                    }
                    return this._delegate.setElementAttribute(renderElement, attributeName, attributeValue);
                }
                /**
                 * Used only in debug mode to serialize property changes to comment nodes,
                 * such as <template> placeholders.
                 */
                setBindingDebugInfo(renderElement, propertyName, propertyValue) {
                    return this._delegate.setBindingDebugInfo(renderElement, propertyName, propertyValue);
                }
                /**
                 * Used only in development mode to set information needed by the DebugNode for this element.
                 */
                setElementDebugInfo(renderElement, info) {
                    var debugEl = debug_node_1.getDebugNode(renderElement);
                    debugEl.setDebugInfo(info);
                    return this._delegate.setElementDebugInfo(renderElement, info);
                }
                setElementClass(renderElement, className, isAdd) {
                    return this._delegate.setElementClass(renderElement, className, isAdd);
                }
                setElementStyle(renderElement, styleName, styleValue) {
                    return this._delegate.setElementStyle(renderElement, styleName, styleValue);
                }
                invokeElementMethod(renderElement, methodName, args) {
                    return this._delegate.invokeElementMethod(renderElement, methodName, args);
                }
                setText(renderNode, text) { return this._delegate.setText(renderNode, text); }
            }
            exports_1("DebugDomRenderer", DebugDomRenderer);
        }
    }
});
