// angular2/ts/src/facade/promise.js
System.register([], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var PromiseWrapper;
    return {
        setters:[],
        execute: function() {
            class PromiseWrapper {
                static resolve(obj) { return Promise.resolve(obj); }
                static reject(obj, _) { return Promise.reject(obj); }
                // Note: We can't rename this method into `catch`, as this is not a valid
                // method name in Dart.
                static catchError(promise, onError) {
                    return promise.catch(onError);
                }
                static all(promises) {
                    if (promises.length == 0)
                        return Promise.resolve([]);
                    return Promise.all(promises);
                }
                static then(promise, success, rejection) {
                    return promise.then(success, rejection);
                }
                static wrap(computation) {
                    return new Promise((res, rej) => {
                        try {
                            res(computation());
                        }
                        catch (e) {
                            rej(e);
                        }
                    });
                }
                static scheduleMicrotask(computation) {
                    PromiseWrapper.then(PromiseWrapper.resolve(null), computation, (_) => { });
                }
                static isPromise(obj) { return obj instanceof Promise; }
                static completer() {
                    var resolve;
                    var reject;
                    var p = new Promise(function (res, rej) {
                        resolve = res;
                        reject = rej;
                    });
                    return { promise: p, resolve: resolve, reject: reject };
                }
            }
            exports_1("PromiseWrapper", PromiseWrapper);
        }
    }
});
