// angular2/ts/src/compiler/html_ast.js
System.register(['angular2/src/facade/lang'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var lang_1;
    var HtmlTextAst, HtmlAttrAst, HtmlElementAst;
    function htmlVisitAll(visitor, asts, context = null) {
        var result = [];
        asts.forEach(ast => {
            var astResult = ast.visit(visitor, context);
            if (lang_1.isPresent(astResult)) {
                result.push(astResult);
            }
        });
        return result;
    }
    exports_1("htmlVisitAll", htmlVisitAll);
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            }],
        execute: function() {
            class HtmlTextAst {
                constructor(value, sourceSpan) {
                    this.value = value;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) { return visitor.visitText(this, context); }
            }
            exports_1("HtmlTextAst", HtmlTextAst);
            class HtmlAttrAst {
                constructor(name, value, sourceSpan) {
                    this.name = name;
                    this.value = value;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) { return visitor.visitAttr(this, context); }
            }
            exports_1("HtmlAttrAst", HtmlAttrAst);
            class HtmlElementAst {
                constructor(name, attrs, children, sourceSpan) {
                    this.name = name;
                    this.attrs = attrs;
                    this.children = children;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) { return visitor.visitElement(this, context); }
            }
            exports_1("HtmlElementAst", HtmlElementAst);
        }
    }
});
