// angular2/ts/src/compiler/source_module.js
System.register(['angular2/src/facade/lang'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var lang_1;
    var MODULE_REGEXP, SourceModule, SourceExpression, SourceExpressions, SourceWithImports;
    function moduleRef(moduleUrl) {
        return `#MODULE[${moduleUrl}]`;
    }
    exports_1("moduleRef", moduleRef);
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            }],
        execute: function() {
            MODULE_REGEXP = /#MODULE\[([^\]]*)\]/g;
            /**
             * Represents generated source code with module references. Internal to the Angular compiler.
             */
            class SourceModule {
                constructor(moduleUrl, sourceWithModuleRefs) {
                    this.moduleUrl = moduleUrl;
                    this.sourceWithModuleRefs = sourceWithModuleRefs;
                }
                static getSourceWithoutImports(sourceWithModuleRefs) {
                    return lang_1.StringWrapper.replaceAllMapped(sourceWithModuleRefs, MODULE_REGEXP, (match) => '');
                }
                getSourceWithImports() {
                    var moduleAliases = {};
                    var imports = [];
                    var newSource = lang_1.StringWrapper.replaceAllMapped(this.sourceWithModuleRefs, MODULE_REGEXP, (match) => {
                        var moduleUrl = match[1];
                        var alias = moduleAliases[moduleUrl];
                        if (lang_1.isBlank(alias)) {
                            if (moduleUrl == this.moduleUrl) {
                                alias = '';
                            }
                            else {
                                alias = `import${imports.length}`;
                                imports.push([moduleUrl, alias]);
                            }
                            moduleAliases[moduleUrl] = alias;
                        }
                        return alias.length > 0 ? `${alias}.` : '';
                    });
                    return new SourceWithImports(newSource, imports);
                }
            }
            exports_1("SourceModule", SourceModule);
            class SourceExpression {
                constructor(declarations, expression) {
                    this.declarations = declarations;
                    this.expression = expression;
                }
            }
            exports_1("SourceExpression", SourceExpression);
            class SourceExpressions {
                constructor(declarations, expressions) {
                    this.declarations = declarations;
                    this.expressions = expressions;
                }
            }
            exports_1("SourceExpressions", SourceExpressions);
            /**
             * Represents generated source code with imports. Internal to the Angular compiler.
             */
            class SourceWithImports {
                constructor(source, imports) {
                    this.source = source;
                    this.imports = imports;
                }
            }
            exports_1("SourceWithImports", SourceWithImports);
        }
    }
});
