// angular2/ts/src/compiler/template_ast.js
System.register(['angular2/src/facade/lang'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var lang_1;
    var TextAst, BoundTextAst, AttrAst, BoundElementPropertyAst, BoundEventAst, VariableAst, ElementAst, EmbeddedTemplateAst, BoundDirectivePropertyAst, DirectiveAst, NgContentAst, PropertyBindingType;
    /**
     * Visit every node in a list of {@link TemplateAst}s with the given {@link TemplateAstVisitor}.
     */
    function templateVisitAll(visitor, asts, context = null) {
        var result = [];
        asts.forEach(ast => {
            var astResult = ast.visit(visitor, context);
            if (lang_1.isPresent(astResult)) {
                result.push(astResult);
            }
        });
        return result;
    }
    exports_1("templateVisitAll", templateVisitAll);
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            }],
        execute: function() {
            /**
             * A segment of text within the template.
             */
            class TextAst {
                constructor(value, ngContentIndex, sourceSpan) {
                    this.value = value;
                    this.ngContentIndex = ngContentIndex;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) { return visitor.visitText(this, context); }
            }
            exports_1("TextAst", TextAst);
            /**
             * A bound expression within the text of a template.
             */
            class BoundTextAst {
                constructor(value, ngContentIndex, sourceSpan) {
                    this.value = value;
                    this.ngContentIndex = ngContentIndex;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) {
                    return visitor.visitBoundText(this, context);
                }
            }
            exports_1("BoundTextAst", BoundTextAst);
            /**
             * A plain attribute on an element.
             */
            class AttrAst {
                constructor(name, value, sourceSpan) {
                    this.name = name;
                    this.value = value;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) { return visitor.visitAttr(this, context); }
            }
            exports_1("AttrAst", AttrAst);
            /**
             * A binding for an element property (e.g. `[property]="expression"`).
             */
            class BoundElementPropertyAst {
                constructor(name, type, value, unit, sourceSpan) {
                    this.name = name;
                    this.type = type;
                    this.value = value;
                    this.unit = unit;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) {
                    return visitor.visitElementProperty(this, context);
                }
            }
            exports_1("BoundElementPropertyAst", BoundElementPropertyAst);
            /**
             * A binding for an element event (e.g. `(event)="handler()"`).
             */
            class BoundEventAst {
                constructor(name, target, handler, sourceSpan) {
                    this.name = name;
                    this.target = target;
                    this.handler = handler;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) {
                    return visitor.visitEvent(this, context);
                }
                get fullName() {
                    if (lang_1.isPresent(this.target)) {
                        return `${this.target}:${this.name}`;
                    }
                    else {
                        return this.name;
                    }
                }
            }
            exports_1("BoundEventAst", BoundEventAst);
            /**
             * A variable declaration on an element (e.g. `#var="expression"`).
             */
            class VariableAst {
                constructor(name, value, sourceSpan) {
                    this.name = name;
                    this.value = value;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) {
                    return visitor.visitVariable(this, context);
                }
            }
            exports_1("VariableAst", VariableAst);
            /**
             * An element declaration in a template.
             */
            class ElementAst {
                constructor(name, attrs, inputs, outputs, exportAsVars, directives, children, ngContentIndex, sourceSpan) {
                    this.name = name;
                    this.attrs = attrs;
                    this.inputs = inputs;
                    this.outputs = outputs;
                    this.exportAsVars = exportAsVars;
                    this.directives = directives;
                    this.children = children;
                    this.ngContentIndex = ngContentIndex;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) {
                    return visitor.visitElement(this, context);
                }
                /**
                 * Whether the element has any active bindings (inputs, outputs, vars, or directives).
                 */
                isBound() {
                    return (this.inputs.length > 0 || this.outputs.length > 0 || this.exportAsVars.length > 0 ||
                        this.directives.length > 0);
                }
                /**
                 * Get the component associated with this element, if any.
                 */
                getComponent() {
                    return this.directives.length > 0 && this.directives[0].directive.isComponent ?
                        this.directives[0].directive :
                        null;
                }
            }
            exports_1("ElementAst", ElementAst);
            /**
             * A `<template>` element included in an Angular template.
             */
            class EmbeddedTemplateAst {
                constructor(attrs, outputs, vars, directives, children, ngContentIndex, sourceSpan) {
                    this.attrs = attrs;
                    this.outputs = outputs;
                    this.vars = vars;
                    this.directives = directives;
                    this.children = children;
                    this.ngContentIndex = ngContentIndex;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) {
                    return visitor.visitEmbeddedTemplate(this, context);
                }
            }
            exports_1("EmbeddedTemplateAst", EmbeddedTemplateAst);
            /**
             * A directive property with a bound value (e.g. `*ngIf="condition").
             */
            class BoundDirectivePropertyAst {
                constructor(directiveName, templateName, value, sourceSpan) {
                    this.directiveName = directiveName;
                    this.templateName = templateName;
                    this.value = value;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) {
                    return visitor.visitDirectiveProperty(this, context);
                }
            }
            exports_1("BoundDirectivePropertyAst", BoundDirectivePropertyAst);
            /**
             * A directive declared on an element.
             */
            class DirectiveAst {
                constructor(directive, inputs, hostProperties, hostEvents, exportAsVars, sourceSpan) {
                    this.directive = directive;
                    this.inputs = inputs;
                    this.hostProperties = hostProperties;
                    this.hostEvents = hostEvents;
                    this.exportAsVars = exportAsVars;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) {
                    return visitor.visitDirective(this, context);
                }
            }
            exports_1("DirectiveAst", DirectiveAst);
            /**
             * Position where content is to be projected (instance of `<ng-content>` in a template).
             */
            class NgContentAst {
                constructor(index, ngContentIndex, sourceSpan) {
                    this.index = index;
                    this.ngContentIndex = ngContentIndex;
                    this.sourceSpan = sourceSpan;
                }
                visit(visitor, context) {
                    return visitor.visitNgContent(this, context);
                }
            }
            exports_1("NgContentAst", NgContentAst);
            /**
             * Enumeration of types of property bindings.
             */
            (function (PropertyBindingType) {
                /**
                 * A normal binding to a property (e.g. `[property]="expression"`).
                 */
                PropertyBindingType[PropertyBindingType["Property"] = 0] = "Property";
                /**
                 * A binding to an element attribute (e.g. `[attr.name]="expression"`).
                 */
                PropertyBindingType[PropertyBindingType["Attribute"] = 1] = "Attribute";
                /**
                 * A binding to a CSS class (e.g. `[class.name]="condition"`).
                 */
                PropertyBindingType[PropertyBindingType["Class"] = 2] = "Class";
                /**
                 * A binding to a style rule (e.g. `[style.rule]="expression"`).
                 */
                PropertyBindingType[PropertyBindingType["Style"] = 3] = "Style";
            })(PropertyBindingType || (PropertyBindingType = {}));
            exports_1("PropertyBindingType", PropertyBindingType);
        }
    }
});
