// angular2/ts/src/compiler/template_normalizer.js
System.register(['./directive_metadata', 'angular2/src/facade/lang', 'angular2/src/facade/exceptions', 'angular2/src/facade/async', 'angular2/src/compiler/xhr', 'angular2/src/compiler/url_resolver', './style_url_resolver', 'angular2/src/core/di', 'angular2/src/core/metadata/view', './html_ast', './html_parser', './template_preparser'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var directive_metadata_1, lang_1, exceptions_1, async_1, xhr_1, url_resolver_1, style_url_resolver_1, di_1, view_1, html_ast_1, html_parser_1, template_preparser_1;
    var TemplateNormalizer, TemplatePreparseVisitor;
    return {
        setters:[
            function (directive_metadata_1_1) {
                directive_metadata_1 = directive_metadata_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (exceptions_1_1) {
                exceptions_1 = exceptions_1_1;
            },
            function (async_1_1) {
                async_1 = async_1_1;
            },
            function (xhr_1_1) {
                xhr_1 = xhr_1_1;
            },
            function (url_resolver_1_1) {
                url_resolver_1 = url_resolver_1_1;
            },
            function (style_url_resolver_1_1) {
                style_url_resolver_1 = style_url_resolver_1_1;
            },
            function (di_1_1) {
                di_1 = di_1_1;
            },
            function (view_1_1) {
                view_1 = view_1_1;
            },
            function (html_ast_1_1) {
                html_ast_1 = html_ast_1_1;
            },
            function (html_parser_1_1) {
                html_parser_1 = html_parser_1_1;
            },
            function (template_preparser_1_1) {
                template_preparser_1 = template_preparser_1_1;
            }],
        execute: function() {
            let TemplateNormalizer = class TemplateNormalizer {
                constructor(_xhr, _urlResolver, _htmlParser) {
                    this._xhr = _xhr;
                    this._urlResolver = _urlResolver;
                    this._htmlParser = _htmlParser;
                }
                normalizeTemplate(directiveType, template) {
                    if (lang_1.isPresent(template.template)) {
                        return async_1.PromiseWrapper.resolve(this.normalizeLoadedTemplate(directiveType, template, template.template, directiveType.moduleUrl));
                    }
                    else if (lang_1.isPresent(template.templateUrl)) {
                        var sourceAbsUrl = this._urlResolver.resolve(directiveType.moduleUrl, template.templateUrl);
                        return this._xhr.get(sourceAbsUrl)
                            .then(templateContent => this.normalizeLoadedTemplate(directiveType, template, templateContent, sourceAbsUrl));
                    }
                    else {
                        throw new exceptions_1.BaseException(`No template specified for component ${directiveType.name}`);
                    }
                }
                normalizeLoadedTemplate(directiveType, templateMeta, template, templateAbsUrl) {
                    var rootNodesAndErrors = this._htmlParser.parse(template, directiveType.name);
                    if (rootNodesAndErrors.errors.length > 0) {
                        var errorString = rootNodesAndErrors.errors.join('\n');
                        throw new exceptions_1.BaseException(`Template parse errors:\n${errorString}`);
                    }
                    var visitor = new TemplatePreparseVisitor();
                    html_ast_1.htmlVisitAll(visitor, rootNodesAndErrors.rootNodes);
                    var allStyles = templateMeta.styles.concat(visitor.styles);
                    var allStyleAbsUrls = visitor.styleUrls.filter(style_url_resolver_1.isStyleUrlResolvable)
                        .map(url => this._urlResolver.resolve(templateAbsUrl, url))
                        .concat(templateMeta.styleUrls.filter(style_url_resolver_1.isStyleUrlResolvable)
                        .map(url => this._urlResolver.resolve(directiveType.moduleUrl, url)));
                    var allResolvedStyles = allStyles.map(style => {
                        var styleWithImports = style_url_resolver_1.extractStyleUrls(this._urlResolver, templateAbsUrl, style);
                        styleWithImports.styleUrls.forEach(styleUrl => allStyleAbsUrls.push(styleUrl));
                        return styleWithImports.style;
                    });
                    var encapsulation = templateMeta.encapsulation;
                    if (encapsulation === view_1.ViewEncapsulation.Emulated && allResolvedStyles.length === 0 &&
                        allStyleAbsUrls.length === 0) {
                        encapsulation = view_1.ViewEncapsulation.None;
                    }
                    return new directive_metadata_1.CompileTemplateMetadata({
                        encapsulation: encapsulation,
                        template: template,
                        templateUrl: templateAbsUrl,
                        styles: allResolvedStyles,
                        styleUrls: allStyleAbsUrls,
                        ngContentSelectors: visitor.ngContentSelectors
                    });
                }
            };
            TemplateNormalizer = __decorate([
                di_1.Injectable(), 
                __metadata('design:paramtypes', [xhr_1.XHR, url_resolver_1.UrlResolver, html_parser_1.HtmlParser])
            ], TemplateNormalizer);
            exports_1("TemplateNormalizer", TemplateNormalizer);
            class TemplatePreparseVisitor {
                constructor() {
                    this.ngContentSelectors = [];
                    this.styles = [];
                    this.styleUrls = [];
                    this.ngNonBindableStackCount = 0;
                }
                visitElement(ast, context) {
                    var preparsedElement = template_preparser_1.preparseElement(ast);
                    switch (preparsedElement.type) {
                        case template_preparser_1.PreparsedElementType.NG_CONTENT:
                            if (this.ngNonBindableStackCount === 0) {
                                this.ngContentSelectors.push(preparsedElement.selectAttr);
                            }
                            break;
                        case template_preparser_1.PreparsedElementType.STYLE:
                            var textContent = '';
                            ast.children.forEach(child => {
                                if (child instanceof html_ast_1.HtmlTextAst) {
                                    textContent += child.value;
                                }
                            });
                            this.styles.push(textContent);
                            break;
                        case template_preparser_1.PreparsedElementType.STYLESHEET:
                            this.styleUrls.push(preparsedElement.hrefAttr);
                            break;
                        default:
                            // DDC reports this as error. See:
                            // https://github.com/dart-lang/dev_compiler/issues/428
                            break;
                    }
                    if (preparsedElement.nonBindable) {
                        this.ngNonBindableStackCount++;
                    }
                    html_ast_1.htmlVisitAll(this, ast.children);
                    if (preparsedElement.nonBindable) {
                        this.ngNonBindableStackCount--;
                    }
                    return null;
                }
                visitAttr(ast, context) { return null; }
                visitText(ast, context) { return null; }
            }
        }
    }
});
