// angular2/ts/src/web_workers/shared/post_message_bus.js
System.register(['angular2/src/facade/exceptions', 'angular2/src/facade/async', 'angular2/src/facade/collection', "angular2/src/core/di"], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var exceptions_1, async_1, collection_1, di_1;
    var PostMessageBus, PostMessageBusSink, PostMessageBusSource, _Channel;
    return {
        setters:[
            function (exceptions_1_1) {
                exceptions_1 = exceptions_1_1;
            },
            function (async_1_1) {
                async_1 = async_1_1;
            },
            function (collection_1_1) {
                collection_1 = collection_1_1;
            },
            function (di_1_1) {
                di_1 = di_1_1;
            }],
        execute: function() {
            /**
             * A TypeScript implementation of {@link MessageBus} for communicating via JavaScript's
             * postMessage API.
             */
            let PostMessageBus = class PostMessageBus {
                constructor(sink, source) {
                    this.sink = sink;
                    this.source = source;
                }
                attachToZone(zone) {
                    this.source.attachToZone(zone);
                    this.sink.attachToZone(zone);
                }
                initChannel(channel, runInZone = true) {
                    this.source.initChannel(channel, runInZone);
                    this.sink.initChannel(channel, runInZone);
                }
                from(channel) { return this.source.from(channel); }
                to(channel) { return this.sink.to(channel); }
            };
            PostMessageBus = __decorate([
                di_1.Injectable(), 
                __metadata('design:paramtypes', [PostMessageBusSink, PostMessageBusSource])
            ], PostMessageBus);
            exports_1("PostMessageBus", PostMessageBus);
            class PostMessageBusSink {
                constructor(_postMessageTarget) {
                    this._postMessageTarget = _postMessageTarget;
                    this._channels = collection_1.StringMapWrapper.create();
                    this._messageBuffer = [];
                }
                attachToZone(zone) {
                    this._zone = zone;
                    this._zone.runOutsideAngular(() => {
                        async_1.ObservableWrapper.subscribe(this._zone.onEventDone, (_) => { this._handleOnEventDone(); });
                    });
                }
                initChannel(channel, runInZone = true) {
                    if (collection_1.StringMapWrapper.contains(this._channels, channel)) {
                        throw new exceptions_1.BaseException(`${channel} has already been initialized`);
                    }
                    var emitter = new async_1.EventEmitter();
                    var channelInfo = new _Channel(emitter, runInZone);
                    this._channels[channel] = channelInfo;
                    emitter.subscribe((data) => {
                        var message = { channel: channel, message: data };
                        if (runInZone) {
                            this._messageBuffer.push(message);
                        }
                        else {
                            this._sendMessages([message]);
                        }
                    });
                }
                to(channel) {
                    if (collection_1.StringMapWrapper.contains(this._channels, channel)) {
                        return this._channels[channel].emitter;
                    }
                    else {
                        throw new exceptions_1.BaseException(`${channel} is not set up. Did you forget to call initChannel?`);
                    }
                }
                _handleOnEventDone() {
                    if (this._messageBuffer.length > 0) {
                        this._sendMessages(this._messageBuffer);
                        this._messageBuffer = [];
                    }
                }
                _sendMessages(messages) { this._postMessageTarget.postMessage(messages); }
            }
            exports_1("PostMessageBusSink", PostMessageBusSink);
            class PostMessageBusSource {
                constructor(eventTarget) {
                    this._channels = collection_1.StringMapWrapper.create();
                    if (eventTarget) {
                        eventTarget.addEventListener("message", (ev) => this._handleMessages(ev));
                    }
                    else {
                        // if no eventTarget is given we assume we're in a WebWorker and listen on the global scope
                        addEventListener("message", (ev) => this._handleMessages(ev));
                    }
                }
                attachToZone(zone) { this._zone = zone; }
                initChannel(channel, runInZone = true) {
                    if (collection_1.StringMapWrapper.contains(this._channels, channel)) {
                        throw new exceptions_1.BaseException(`${channel} has already been initialized`);
                    }
                    var emitter = new async_1.EventEmitter();
                    var channelInfo = new _Channel(emitter, runInZone);
                    this._channels[channel] = channelInfo;
                }
                from(channel) {
                    if (collection_1.StringMapWrapper.contains(this._channels, channel)) {
                        return this._channels[channel].emitter;
                    }
                    else {
                        throw new exceptions_1.BaseException(`${channel} is not set up. Did you forget to call initChannel?`);
                    }
                }
                _handleMessages(ev) {
                    var messages = ev.data;
                    for (var i = 0; i < messages.length; i++) {
                        this._handleMessage(messages[i]);
                    }
                }
                _handleMessage(data) {
                    var channel = data.channel;
                    if (collection_1.StringMapWrapper.contains(this._channels, channel)) {
                        var channelInfo = this._channels[channel];
                        if (channelInfo.runInZone) {
                            this._zone.run(() => { channelInfo.emitter.emit(data.message); });
                        }
                        else {
                            channelInfo.emitter.emit(data.message);
                        }
                    }
                }
            }
            exports_1("PostMessageBusSource", PostMessageBusSource);
            /**
             * Helper class that wraps a channel's {@link EventEmitter} and
             * keeps track of if it should run in the zone.
             */
            class _Channel {
                constructor(emitter, runInZone) {
                    this.emitter = emitter;
                    this.runInZone = runInZone;
                }
            }
        }
    }
});
