// angular2/ts/src/platform/browser/testability.js
System.register(['angular2/src/facade/collection', 'angular2/src/facade/lang', 'angular2/src/platform/dom/dom_adapter', 'angular2/core'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var collection_1, lang_1, dom_adapter_1, core_1;
    var PublicTestability, BrowserGetTestability;
    return {
        setters:[
            function (collection_1_1) {
                collection_1 = collection_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (dom_adapter_1_1) {
                dom_adapter_1 = dom_adapter_1_1;
            },
            function (core_1_1) {
                core_1 = core_1_1;
            }],
        execute: function() {
            class PublicTestability {
                constructor(testability) {
                    this._testability = testability;
                }
                isStable() { return this._testability.isStable(); }
                whenStable(callback) { this._testability.whenStable(callback); }
                findBindings(using, provider, exactMatch) {
                    return this.findProviders(using, provider, exactMatch);
                }
                findProviders(using, provider, exactMatch) {
                    return this._testability.findBindings(using, provider, exactMatch);
                }
            }
            class BrowserGetTestability {
                static init() { core_1.setTestabilityGetter(new BrowserGetTestability()); }
                addToWindow(registry) {
                    lang_1.global.getAngularTestability = (elem, findInAncestors = true) => {
                        var testability = registry.findTestabilityInTree(elem, findInAncestors);
                        if (testability == null) {
                            throw new Error('Could not find testability for element.');
                        }
                        return new PublicTestability(testability);
                    };
                    lang_1.global.getAllAngularTestabilities = () => {
                        var testabilities = registry.getAllTestabilities();
                        return testabilities.map((testability) => { return new PublicTestability(testability); });
                    };
                    var whenAllStable = (callback) => {
                        var testabilities = lang_1.global.getAllAngularTestabilities();
                        var count = testabilities.length;
                        var didWork = false;
                        var decrement = function (didWork_) {
                            didWork = didWork || didWork_;
                            count--;
                            if (count == 0) {
                                callback(didWork);
                            }
                        };
                        testabilities.forEach(function (testability) { testability.whenStable(decrement); });
                    };
                    if (!lang_1.global.frameworkStabilizers) {
                        lang_1.global.frameworkStabilizers = collection_1.ListWrapper.createGrowableSize(0);
                    }
                    lang_1.global.frameworkStabilizers.push(whenAllStable);
                }
                findTestabilityInTree(registry, elem, findInAncestors) {
                    if (elem == null) {
                        return null;
                    }
                    var t = registry.getTestability(elem);
                    if (lang_1.isPresent(t)) {
                        return t;
                    }
                    else if (!findInAncestors) {
                        return null;
                    }
                    if (dom_adapter_1.DOM.isShadowRoot(elem)) {
                        return this.findTestabilityInTree(registry, dom_adapter_1.DOM.getHost(elem), true);
                    }
                    return this.findTestabilityInTree(registry, dom_adapter_1.DOM.parentElement(elem), true);
                }
            }
            exports_1("BrowserGetTestability", BrowserGetTestability);
        }
    }
});
