// angular2/ts/src/http/backends/jsonp_backend.js
System.register(['../interfaces', '../enums', '../static_response', '../base_response_options', 'angular2/core', './browser_jsonp', 'angular2/src/facade/exceptions', 'angular2/src/facade/lang', 'rxjs/Observable'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var interfaces_1, enums_1, static_response_1, base_response_options_1, core_1, browser_jsonp_1, exceptions_1, lang_1, Observable_1;
    var JSONP_ERR_NO_CALLBACK, JSONP_ERR_WRONG_METHOD, JSONPConnection, JSONPConnection_, JSONPBackend, JSONPBackend_;
    return {
        setters:[
            function (interfaces_1_1) {
                interfaces_1 = interfaces_1_1;
            },
            function (enums_1_1) {
                enums_1 = enums_1_1;
            },
            function (static_response_1_1) {
                static_response_1 = static_response_1_1;
            },
            function (base_response_options_1_1) {
                base_response_options_1 = base_response_options_1_1;
            },
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (browser_jsonp_1_1) {
                browser_jsonp_1 = browser_jsonp_1_1;
            },
            function (exceptions_1_1) {
                exceptions_1 = exceptions_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (Observable_1_1) {
                Observable_1 = Observable_1_1;
            }],
        execute: function() {
            JSONP_ERR_NO_CALLBACK = 'JSONP injected script did not invoke callback.';
            JSONP_ERR_WRONG_METHOD = 'JSONP requests must use GET request method.';
            /**
             * Abstract base class for an in-flight JSONP request.
             */
            class JSONPConnection {
            }
            exports_1("JSONPConnection", JSONPConnection);
            class JSONPConnection_ extends JSONPConnection {
                constructor(req, _dom, baseResponseOptions) {
                    super();
                    this._dom = _dom;
                    this.baseResponseOptions = baseResponseOptions;
                    this._finished = false;
                    if (req.method !== enums_1.RequestMethod.Get) {
                        throw exceptions_1.makeTypeError(JSONP_ERR_WRONG_METHOD);
                    }
                    this.request = req;
                    this.response = new Observable_1.Observable((responseObserver) => {
                        this.readyState = enums_1.ReadyState.Loading;
                        let id = this._id = _dom.nextRequestID();
                        _dom.exposeConnection(id, this);
                        // Workaround Dart
                        // url = url.replace(/=JSONP_CALLBACK(&|$)/, `generated method`);
                        let callback = _dom.requestCallback(this._id);
                        let url = req.url;
                        if (url.indexOf('=JSONP_CALLBACK&') > -1) {
                            url = lang_1.StringWrapper.replace(url, '=JSONP_CALLBACK&', `=${callback}&`);
                        }
                        else if (url.lastIndexOf('=JSONP_CALLBACK') === url.length - '=JSONP_CALLBACK'.length) {
                            url = url.substring(0, url.length - '=JSONP_CALLBACK'.length) + `=${callback}`;
                        }
                        let script = this._script = _dom.build(url);
                        let onLoad = (event) => {
                            if (this.readyState === enums_1.ReadyState.Cancelled)
                                return;
                            this.readyState = enums_1.ReadyState.Done;
                            _dom.cleanup(script);
                            if (!this._finished) {
                                let responseOptions = new base_response_options_1.ResponseOptions({ body: JSONP_ERR_NO_CALLBACK, type: enums_1.ResponseType.Error, url: url });
                                if (lang_1.isPresent(baseResponseOptions)) {
                                    responseOptions = baseResponseOptions.merge(responseOptions);
                                }
                                responseObserver.error(new static_response_1.Response(responseOptions));
                                return;
                            }
                            let responseOptions = new base_response_options_1.ResponseOptions({ body: this._responseData, url: url });
                            if (lang_1.isPresent(this.baseResponseOptions)) {
                                responseOptions = this.baseResponseOptions.merge(responseOptions);
                            }
                            responseObserver.next(new static_response_1.Response(responseOptions));
                            responseObserver.complete();
                        };
                        let onError = (error) => {
                            if (this.readyState === enums_1.ReadyState.Cancelled)
                                return;
                            this.readyState = enums_1.ReadyState.Done;
                            _dom.cleanup(script);
                            let responseOptions = new base_response_options_1.ResponseOptions({ body: error.message, type: enums_1.ResponseType.Error });
                            if (lang_1.isPresent(baseResponseOptions)) {
                                responseOptions = baseResponseOptions.merge(responseOptions);
                            }
                            responseObserver.error(new static_response_1.Response(responseOptions));
                        };
                        script.addEventListener('load', onLoad);
                        script.addEventListener('error', onError);
                        _dom.send(script);
                        return () => {
                            this.readyState = enums_1.ReadyState.Cancelled;
                            script.removeEventListener('load', onLoad);
                            script.removeEventListener('error', onError);
                            if (lang_1.isPresent(script)) {
                                this._dom.cleanup(script);
                            }
                        };
                    });
                }
                finished(data) {
                    // Don't leak connections
                    this._finished = true;
                    this._dom.removeConnection(this._id);
                    if (this.readyState === enums_1.ReadyState.Cancelled)
                        return;
                    this._responseData = data;
                }
            }
            exports_1("JSONPConnection_", JSONPConnection_);
            /**
             * A {@link ConnectionBackend} that uses the JSONP strategy of making requests.
             */
            class JSONPBackend extends interfaces_1.ConnectionBackend {
            }
            exports_1("JSONPBackend", JSONPBackend);
            let JSONPBackend_ = class JSONPBackend_ extends JSONPBackend {
                constructor(_browserJSONP, _baseResponseOptions) {
                    super();
                    this._browserJSONP = _browserJSONP;
                    this._baseResponseOptions = _baseResponseOptions;
                }
                createConnection(request) {
                    return new JSONPConnection_(request, this._browserJSONP, this._baseResponseOptions);
                }
            };
            JSONPBackend_ = __decorate([
                core_1.Injectable(), 
                __metadata('design:paramtypes', [browser_jsonp_1.BrowserJsonp, base_response_options_1.ResponseOptions])
            ], JSONPBackend_);
            exports_1("JSONPBackend_", JSONPBackend_);
        }
    }
});
