// angular2/ts/src/platform/dom/dom_renderer.js
System.register(['angular2/src/core/di', 'angular2/src/animate/animation_builder', 'angular2/src/facade/lang', 'angular2/src/facade/exceptions', './shared_styles_host', './events/event_manager', './dom_tokens', 'angular2/src/core/metadata', 'angular2/src/platform/dom/dom_adapter', './util'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __param = (this && this.__param) || function (paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    };
    var di_1, animation_builder_1, lang_1, exceptions_1, shared_styles_host_1, event_manager_1, dom_tokens_1, metadata_1, dom_adapter_1, util_1;
    var NAMESPACE_URIS, TEMPLATE_COMMENT_TEXT, TEMPLATE_BINDINGS_EXP, DomRootRenderer, DomRootRenderer_, DomRenderer, COMPONENT_REGEX, COMPONENT_VARIABLE, HOST_ATTR, CONTENT_ATTR, NS_PREFIX_RE;
    function moveNodesAfterSibling(sibling, nodes) {
        var parent = dom_adapter_1.DOM.parentElement(sibling);
        if (nodes.length > 0 && lang_1.isPresent(parent)) {
            var nextSibling = dom_adapter_1.DOM.nextSibling(sibling);
            if (lang_1.isPresent(nextSibling)) {
                for (var i = 0; i < nodes.length; i++) {
                    dom_adapter_1.DOM.insertBefore(nextSibling, nodes[i]);
                }
            }
            else {
                for (var i = 0; i < nodes.length; i++) {
                    dom_adapter_1.DOM.appendChild(parent, nodes[i]);
                }
            }
        }
    }
    function appendNodes(parent, nodes) {
        for (var i = 0; i < nodes.length; i++) {
            dom_adapter_1.DOM.appendChild(parent, nodes[i]);
        }
    }
    function decoratePreventDefault(eventHandler) {
        return (event) => {
            var allowDefaultBehavior = eventHandler(event);
            if (allowDefaultBehavior === false) {
                // TODO(tbosch): move preventDefault into event plugins...
                dom_adapter_1.DOM.preventDefault(event);
            }
        };
    }
    function _shimContentAttribute(componentShortId) {
        return lang_1.StringWrapper.replaceAll(CONTENT_ATTR, COMPONENT_REGEX, componentShortId);
    }
    function _shimHostAttribute(componentShortId) {
        return lang_1.StringWrapper.replaceAll(HOST_ATTR, COMPONENT_REGEX, componentShortId);
    }
    function _flattenStyles(compId, styles, target) {
        for (var i = 0; i < styles.length; i++) {
            var style = styles[i];
            if (lang_1.isArray(style)) {
                _flattenStyles(compId, style, target);
            }
            else {
                style = lang_1.StringWrapper.replaceAll(style, COMPONENT_REGEX, compId);
                target.push(style);
            }
        }
        return target;
    }
    function splitNamespace(name) {
        if (name[0] != '@') {
            return [null, name];
        }
        let match = lang_1.RegExpWrapper.firstMatch(NS_PREFIX_RE, name);
        return [match[1], match[2]];
    }
    return {
        setters:[
            function (di_1_1) {
                di_1 = di_1_1;
            },
            function (animation_builder_1_1) {
                animation_builder_1 = animation_builder_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (exceptions_1_1) {
                exceptions_1 = exceptions_1_1;
            },
            function (shared_styles_host_1_1) {
                shared_styles_host_1 = shared_styles_host_1_1;
            },
            function (event_manager_1_1) {
                event_manager_1 = event_manager_1_1;
            },
            function (dom_tokens_1_1) {
                dom_tokens_1 = dom_tokens_1_1;
            },
            function (metadata_1_1) {
                metadata_1 = metadata_1_1;
            },
            function (dom_adapter_1_1) {
                dom_adapter_1 = dom_adapter_1_1;
            },
            function (util_1_1) {
                util_1 = util_1_1;
            }],
        execute: function() {
            NAMESPACE_URIS = lang_1.CONST_EXPR({ 'xlink': 'http://www.w3.org/1999/xlink', 'svg': 'http://www.w3.org/2000/svg' });
            TEMPLATE_COMMENT_TEXT = 'template bindings={}';
            TEMPLATE_BINDINGS_EXP = /^template bindings=(.*)$/g;
            class DomRootRenderer {
                constructor(document, eventManager, sharedStylesHost, animate) {
                    this.document = document;
                    this.eventManager = eventManager;
                    this.sharedStylesHost = sharedStylesHost;
                    this.animate = animate;
                    this._registeredComponents = new Map();
                }
                renderComponent(componentProto) {
                    var renderer = this._registeredComponents.get(componentProto.id);
                    if (lang_1.isBlank(renderer)) {
                        renderer = new DomRenderer(this, componentProto);
                        this._registeredComponents.set(componentProto.id, renderer);
                    }
                    return renderer;
                }
            }
            exports_1("DomRootRenderer", DomRootRenderer);
            let DomRootRenderer_ = class DomRootRenderer_ extends DomRootRenderer {
                constructor(_document, _eventManager, sharedStylesHost, animate) {
                    super(_document, _eventManager, sharedStylesHost, animate);
                }
            };
            DomRootRenderer_ = __decorate([
                di_1.Injectable(),
                __param(0, di_1.Inject(dom_tokens_1.DOCUMENT)), 
                __metadata('design:paramtypes', [Object, event_manager_1.EventManager, shared_styles_host_1.DomSharedStylesHost, animation_builder_1.AnimationBuilder])
            ], DomRootRenderer_);
            exports_1("DomRootRenderer_", DomRootRenderer_);
            class DomRenderer {
                constructor(_rootRenderer, componentProto) {
                    this._rootRenderer = _rootRenderer;
                    this.componentProto = componentProto;
                    this._styles = _flattenStyles(componentProto.id, componentProto.styles, []);
                    if (componentProto.encapsulation !== metadata_1.ViewEncapsulation.Native) {
                        this._rootRenderer.sharedStylesHost.addStyles(this._styles);
                    }
                    if (this.componentProto.encapsulation === metadata_1.ViewEncapsulation.Emulated) {
                        this._contentAttr = _shimContentAttribute(componentProto.id);
                        this._hostAttr = _shimHostAttribute(componentProto.id);
                    }
                    else {
                        this._contentAttr = null;
                        this._hostAttr = null;
                    }
                }
                renderComponent(componentProto) {
                    return this._rootRenderer.renderComponent(componentProto);
                }
                selectRootElement(selector) {
                    var el = dom_adapter_1.DOM.querySelector(this._rootRenderer.document, selector);
                    if (lang_1.isBlank(el)) {
                        throw new exceptions_1.BaseException(`The selector "${selector}" did not match any elements`);
                    }
                    dom_adapter_1.DOM.clearNodes(el);
                    return el;
                }
                createElement(parent, name) {
                    var nsAndName = splitNamespace(name);
                    var el = lang_1.isPresent(nsAndName[0]) ?
                        dom_adapter_1.DOM.createElementNS(NAMESPACE_URIS[nsAndName[0]], nsAndName[1]) :
                        dom_adapter_1.DOM.createElement(nsAndName[1]);
                    if (lang_1.isPresent(this._contentAttr)) {
                        dom_adapter_1.DOM.setAttribute(el, this._contentAttr, '');
                    }
                    if (lang_1.isPresent(parent)) {
                        dom_adapter_1.DOM.appendChild(parent, el);
                    }
                    return el;
                }
                createViewRoot(hostElement) {
                    var nodesParent;
                    if (this.componentProto.encapsulation === metadata_1.ViewEncapsulation.Native) {
                        nodesParent = dom_adapter_1.DOM.createShadowRoot(hostElement);
                        this._rootRenderer.sharedStylesHost.addHost(nodesParent);
                        for (var i = 0; i < this._styles.length; i++) {
                            dom_adapter_1.DOM.appendChild(nodesParent, dom_adapter_1.DOM.createStyleElement(this._styles[i]));
                        }
                    }
                    else {
                        if (lang_1.isPresent(this._hostAttr)) {
                            dom_adapter_1.DOM.setAttribute(hostElement, this._hostAttr, '');
                        }
                        nodesParent = hostElement;
                    }
                    return nodesParent;
                }
                createTemplateAnchor(parentElement) {
                    var comment = dom_adapter_1.DOM.createComment(TEMPLATE_COMMENT_TEXT);
                    if (lang_1.isPresent(parentElement)) {
                        dom_adapter_1.DOM.appendChild(parentElement, comment);
                    }
                    return comment;
                }
                createText(parentElement, value) {
                    var node = dom_adapter_1.DOM.createTextNode(value);
                    if (lang_1.isPresent(parentElement)) {
                        dom_adapter_1.DOM.appendChild(parentElement, node);
                    }
                    return node;
                }
                projectNodes(parentElement, nodes) {
                    if (lang_1.isBlank(parentElement))
                        return;
                    appendNodes(parentElement, nodes);
                }
                attachViewAfter(node, viewRootNodes) {
                    moveNodesAfterSibling(node, viewRootNodes);
                    for (let i = 0; i < viewRootNodes.length; i++)
                        this.animateNodeEnter(viewRootNodes[i]);
                }
                detachView(viewRootNodes) {
                    for (var i = 0; i < viewRootNodes.length; i++) {
                        var node = viewRootNodes[i];
                        dom_adapter_1.DOM.remove(node);
                        this.animateNodeLeave(node);
                    }
                }
                destroyView(hostElement, viewAllNodes) {
                    if (this.componentProto.encapsulation === metadata_1.ViewEncapsulation.Native && lang_1.isPresent(hostElement)) {
                        this._rootRenderer.sharedStylesHost.removeHost(dom_adapter_1.DOM.getShadowRoot(hostElement));
                    }
                }
                listen(renderElement, name, callback) {
                    return this._rootRenderer.eventManager.addEventListener(renderElement, name, decoratePreventDefault(callback));
                }
                listenGlobal(target, name, callback) {
                    return this._rootRenderer.eventManager.addGlobalEventListener(target, name, decoratePreventDefault(callback));
                }
                setElementProperty(renderElement, propertyName, propertyValue) {
                    dom_adapter_1.DOM.setProperty(renderElement, propertyName, propertyValue);
                }
                setElementAttribute(renderElement, attributeName, attributeValue) {
                    var attrNs;
                    var nsAndName = splitNamespace(attributeName);
                    if (lang_1.isPresent(nsAndName[0])) {
                        attributeName = nsAndName[0] + ':' + nsAndName[1];
                        attrNs = NAMESPACE_URIS[nsAndName[0]];
                    }
                    if (lang_1.isPresent(attributeValue)) {
                        if (lang_1.isPresent(attrNs)) {
                            dom_adapter_1.DOM.setAttributeNS(renderElement, attrNs, attributeName, attributeValue);
                        }
                        else {
                            dom_adapter_1.DOM.setAttribute(renderElement, nsAndName[1], attributeValue);
                        }
                    }
                    else {
                        dom_adapter_1.DOM.removeAttribute(renderElement, attributeName);
                    }
                }
                setBindingDebugInfo(renderElement, propertyName, propertyValue) {
                    var dashCasedPropertyName = util_1.camelCaseToDashCase(propertyName);
                    if (dom_adapter_1.DOM.isCommentNode(renderElement)) {
                        var existingBindings = lang_1.RegExpWrapper.firstMatch(TEMPLATE_BINDINGS_EXP, lang_1.StringWrapper.replaceAll(dom_adapter_1.DOM.getText(renderElement), /\n/g, ''));
                        var parsedBindings = lang_1.Json.parse(existingBindings[1]);
                        parsedBindings[dashCasedPropertyName] = propertyValue;
                        dom_adapter_1.DOM.setText(renderElement, lang_1.StringWrapper.replace(TEMPLATE_COMMENT_TEXT, '{}', lang_1.Json.stringify(parsedBindings)));
                    }
                    else {
                        this.setElementAttribute(renderElement, propertyName, propertyValue);
                    }
                }
                setElementDebugInfo(renderElement, info) { }
                setElementClass(renderElement, className, isAdd) {
                    if (isAdd) {
                        dom_adapter_1.DOM.addClass(renderElement, className);
                    }
                    else {
                        dom_adapter_1.DOM.removeClass(renderElement, className);
                    }
                }
                setElementStyle(renderElement, styleName, styleValue) {
                    if (lang_1.isPresent(styleValue)) {
                        dom_adapter_1.DOM.setStyle(renderElement, styleName, lang_1.stringify(styleValue));
                    }
                    else {
                        dom_adapter_1.DOM.removeStyle(renderElement, styleName);
                    }
                }
                invokeElementMethod(renderElement, methodName, args) {
                    dom_adapter_1.DOM.invoke(renderElement, methodName, args);
                }
                setText(renderNode, text) { dom_adapter_1.DOM.setText(renderNode, text); }
                /**
                 * Performs animations if necessary
                 * @param node
                 */
                animateNodeEnter(node) {
                    if (dom_adapter_1.DOM.isElementNode(node) && dom_adapter_1.DOM.hasClass(node, 'ng-animate')) {
                        dom_adapter_1.DOM.addClass(node, 'ng-enter');
                        this._rootRenderer.animate.css()
                            .addAnimationClass('ng-enter-active')
                            .start(node)
                            .onComplete(() => { dom_adapter_1.DOM.removeClass(node, 'ng-enter'); });
                    }
                }
                /**
                 * If animations are necessary, performs animations then removes the element; otherwise, it just
                 * removes the element.
                 * @param node
                 */
                animateNodeLeave(node) {
                    if (dom_adapter_1.DOM.isElementNode(node) && dom_adapter_1.DOM.hasClass(node, 'ng-animate')) {
                        dom_adapter_1.DOM.addClass(node, 'ng-leave');
                        this._rootRenderer.animate.css()
                            .addAnimationClass('ng-leave-active')
                            .start(node)
                            .onComplete(() => {
                            dom_adapter_1.DOM.removeClass(node, 'ng-leave');
                            dom_adapter_1.DOM.remove(node);
                        });
                    }
                    else {
                        dom_adapter_1.DOM.remove(node);
                    }
                }
            }
            exports_1("DomRenderer", DomRenderer);
            COMPONENT_REGEX = /%COMP%/g;
            exports_1("COMPONENT_VARIABLE", COMPONENT_VARIABLE = '%COMP%');
            exports_1("HOST_ATTR", HOST_ATTR = `_nghost-${COMPONENT_VARIABLE}`);
            exports_1("CONTENT_ATTR", CONTENT_ATTR = `_ngcontent-${COMPONENT_VARIABLE}`);
            NS_PREFIX_RE = /^@([^:]+):(.+)/g;
        }
    }
});
