// angular2/ts/src/core/render/api.js
System.register([], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var RenderComponentType, RenderDebugInfo, Renderer, RootRenderer;
    return {
        setters:[],
        execute: function() {
            class RenderComponentType {
                constructor(id, encapsulation, styles) {
                    this.id = id;
                    this.encapsulation = encapsulation;
                    this.styles = styles;
                }
            }
            exports_1("RenderComponentType", RenderComponentType);
            class RenderDebugInfo {
                constructor(injector, component, providerTokens, locals) {
                    this.injector = injector;
                    this.component = component;
                    this.providerTokens = providerTokens;
                    this.locals = locals;
                }
            }
            exports_1("RenderDebugInfo", RenderDebugInfo);
            class Renderer {
            }
            exports_1("Renderer", Renderer);
            /**
             * Injectable service that provides a low-level interface for modifying the UI.
             *
             * Use this service to bypass Angular's templating and make custom UI changes that can't be
             * expressed declaratively. For example if you need to set a property or an attribute whose name is
             * not statically known, use {@link #setElementProperty} or {@link #setElementAttribute}
             * respectively.
             *
             * If you are implementing a custom renderer, you must implement this interface.
             *
             * The default Renderer implementation is `DomRenderer`. Also available is `WebWorkerRenderer`.
             */
            class RootRenderer {
            }
            exports_1("RootRenderer", RootRenderer);
        }
    }
});
