// angular2/ts/examples/router/ts/can_deactivate/can_deactivate_example.js
System.register(['angular2/core', 'angular2/bootstrap', 'angular2/router'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var core_1, bootstrap_1, router_1;
    var NoteCmp, NoteIndexCmp, AppCmp;
    function main() {
        return bootstrap_1.bootstrap(AppCmp, [core_1.provide(router_1.APP_BASE_HREF, { useValue: '/angular2/examples/router/ts/can_deactivate' })]);
    }
    exports_1("main", main);
    return {
        setters:[
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (bootstrap_1_1) {
                bootstrap_1 = bootstrap_1_1;
            },
            function (router_1_1) {
                router_1 = router_1_1;
            }],
        execute: function() {
            // #docregion routerCanDeactivate
            let NoteCmp = class NoteCmp {
                constructor(params) {
                    this.id = params.get('id');
                }
                routerCanDeactivate(next, prev) {
                    return confirm('Are you sure you want to leave?');
                }
            };
            NoteCmp = __decorate([
                core_1.Component({
                    selector: 'note-cmp',
                    template: `
    <div>
      <h2>id: {{id}}</h2>
      <textarea cols="40" rows="10"></textarea>
    </div>`
                }), 
                __metadata('design:paramtypes', [router_1.RouteParams])
            ], NoteCmp);
            // #enddocregion
            let NoteIndexCmp = class NoteIndexCmp {
            };
            NoteIndexCmp = __decorate([
                core_1.Component({
                    selector: 'note-index-cmp',
                    template: `
    <h1>Your Notes</h1>
    <div>
      Edit <a [routerLink]="['/NoteCmp', {id: 1}]" id="note-1-link">Note 1</a> |
      Edit <a [routerLink]="['/NoteCmp', {id: 2}]" id="note-2-link">Note 2</a>
    </div>
  `,
                    directives: [router_1.ROUTER_DIRECTIVES]
                }), 
                __metadata('design:paramtypes', [])
            ], NoteIndexCmp);
            let AppCmp = class AppCmp {
            };
            AppCmp = __decorate([
                core_1.Component({
                    selector: 'example-app',
                    template: `
    <h1>My App</h1>
    <router-outlet></router-outlet>
  `,
                    directives: [router_1.ROUTER_DIRECTIVES]
                }),
                router_1.RouteConfig([
                    { path: '/note/:id', component: NoteCmp, name: 'NoteCmp' },
                    { path: '/', component: NoteIndexCmp, name: 'NoteIndexCmp' }
                ]), 
                __metadata('design:paramtypes', [])
            ], AppCmp);
        }
    }
});
