// angular2/ts/src/mock/mock_location_strategy.js
System.register(['angular2/src/core/di', 'angular2/src/facade/async', 'angular2/src/router/location_strategy'], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var di_1, async_1, location_strategy_1;
    var MockLocationStrategy, _MockPopStateEvent;
    return {
        setters:[
            function (di_1_1) {
                di_1 = di_1_1;
            },
            function (async_1_1) {
                async_1 = async_1_1;
            },
            function (location_strategy_1_1) {
                location_strategy_1 = location_strategy_1_1;
            }],
        execute: function() {
            /**
             * A mock implementation of {@link LocationStrategy} that allows tests to fire simulated
             * location events.
             */
            let MockLocationStrategy = class MockLocationStrategy extends location_strategy_1.LocationStrategy {
                constructor() {
                    super();
                    this.internalBaseHref = '/';
                    this.internalPath = '/';
                    this.internalTitle = '';
                    this.urlChanges = [];
                    /** @internal */
                    this._subject = new async_1.EventEmitter();
                }
                simulatePopState(url) {
                    this.internalPath = url;
                    async_1.ObservableWrapper.callEmit(this._subject, new _MockPopStateEvent(this.path()));
                }
                path() { return this.internalPath; }
                prepareExternalUrl(internal) {
                    if (internal.startsWith('/') && this.internalBaseHref.endsWith('/')) {
                        return this.internalBaseHref + internal.substring(1);
                    }
                    return this.internalBaseHref + internal;
                }
                pushState(ctx, title, path, query) {
                    this.internalTitle = title;
                    var url = path + (query.length > 0 ? ('?' + query) : '');
                    this.internalPath = url;
                    var externalUrl = this.prepareExternalUrl(url);
                    this.urlChanges.push(externalUrl);
                }
                replaceState(ctx, title, path, query) {
                    this.internalTitle = title;
                    var url = path + (query.length > 0 ? ('?' + query) : '');
                    this.internalPath = url;
                    var externalUrl = this.prepareExternalUrl(url);
                    this.urlChanges.push('replace: ' + externalUrl);
                }
                onPopState(fn) { async_1.ObservableWrapper.subscribe(this._subject, fn); }
                getBaseHref() { return this.internalBaseHref; }
                back() {
                    if (this.urlChanges.length > 0) {
                        this.urlChanges.pop();
                        var nextUrl = this.urlChanges.length > 0 ? this.urlChanges[this.urlChanges.length - 1] : '';
                        this.simulatePopState(nextUrl);
                    }
                }
                forward() { throw 'not implemented'; }
            };
            MockLocationStrategy = __decorate([
                di_1.Injectable(), 
                __metadata('design:paramtypes', [])
            ], MockLocationStrategy);
            exports_1("MockLocationStrategy", MockLocationStrategy);
            class _MockPopStateEvent {
                constructor(newUrl) {
                    this.newUrl = newUrl;
                    this.pop = true;
                    this.type = 'popstate';
                }
            }
        }
    }
});
